docs/aspnetcore-facility.md

```diff
@@ -10,38 +10,78 @@ The ASP.NET Core facility provides Castle Windsor integration using a custom act
-extension. You can opt for your own conventional registrations instead like so:
+`Controllers`, `ViewComponents` and `TagHelpers` are registered automatically for you when the `services.AddCastleWindsor(container)` extension
+is called. All components are registered with a scoped lifestyle, and are assumed to be a single instance for the duration of the web request. 
+You can optionally override the lifestyles for each of these if you prefer using the `options` callback in the example below:
 
 ```csharp
-container.Register(Classes.FromAssemblyInThisApplication(typeof(TStartup).Assembly)
-	.BasedOn<Controller>().LifestyleScoped()); // <- Change lifestyle
-container.Register(Classes.FromAssemblyInThisApplication(typeof(TStartup).Assembly)
-	.BasedOn<ViewComponent>().LifestyleTransient());
-container.Register(Classes.FromAssemblyInThisApplication(typeof(TStartup).Assembly)
-	.BasedOn<TagHelper>().LifestyleTransient());
+services.AddCastleWindsor(container, opts =>
+{
+	opts.RegisterControllers(typeof(HomeController).Assembly, LifestyleType.Transient);
+	opts.RegisterTagHelpers(typeof(EmailTagHelper).Assembly, LifestyleType.Transient);
+	opts.RegisterViewComponents(typeof(AddressComponent).Assembly, LifestyleType.Transient);
+});
+```
+
+This is also useful if your framework components live in a separate assemblies or are not defined in the same assembly as your web application.
+Alternatively if your framework components all live in one assembly and you dont need to change lifestyles then you can simply use the following:
+
+```csharp
+services.AddCastleWindsor(container, opts =>
+{
+	opts.UseEntryAssembly(typeof(HomeController).Assembly);
+});
+```
+This is good for trouble shooting situations where nothing get's registered because of problems in the hosting environment where 
+GetEntryAssembly/GetCallingAssembly does not work as expected. 
+
+### Registering custom middleware
+
+If you don't have any DI requirements in your middleware you can simply register it in the IServiceCollection like so: 
+
+```csharp
+services.AddSingleton<FrameworkMiddleware>(); // Do this if you don't care about using Windsor
+```
+
+Alternatively if you would like to inject dependencies into your middleware that are registered with Castle Windsor then you can use the 
+`UseCastleWindsorMiddleware` extension.
+
+```csharp
+// Add custom middleware, do this if your middleware uses DI from Windsor
+app.UseCastleWindsorMiddleware<CustomMiddleware>(container);
+```
+
+### Cross Wiring into the IServiceCollection
+
+There is an additional feature you can use to `Cross Wire` components into the IServiceCollection. This is useful for cases where the framework
+needs to know how to resolve a type for example with the Razor @Inject directive in MVC projects. 
+
+```csharp
+container.Register(Component.For<IUserService>().ImplementedBy<AspNetUserService>()
+	.CrossWired(container, services, LifestyleType.Scoped));
```

Comment:
It sucks that you have to specify the container in `CrossWired` but I know why.

Reply:
Yup, this sucked hard.

Reply:
Any suggestions?

Reply:
Other than being a little ugly, one problem with this approach is that `CrossWired` registers the service in the ASP.NET Core container before it is even registered in Windsor.

What about using a ComponentModel construction contributor (`IContributeComponentModelConstruction`) that does the cross wiring after the `ComponentModel` is built. The WCF facility has one (`WcfBehaviorInspector`) to set up a custom activator, as do most facilities including typed factory, startable, event wiring, synchronize, and built in things like mixins.

Reply:
> Other than being a little ugly, one problem with this approach is that CrossWired registers the service in the ASP.NET Core container before it is even registered in Windsor.

This raises a really important question about how people use Windsor. I hung this extension off ComponentRegistration<T> because I assumed it would be called within the context of Container.Register(where I might be getting it wrong). The cross wiring here is merely registering late bound components with factories in the IServiceCollection. My assumption was eventually the service will be registered in both places. 

> What about using a ComponentModel construction contributor 

Yes, this would be hard. What CrossWiring is really trying to solve, is making types registered through Windsor available to framework code buried behind IServiceCollection/ServiceProvider logic like the MVC Razor @Inject directive. This is a registration time consideration.  I am worried that construction contribution does not consider the registration requirements and anything you can do during resolve to implement this might already be too late.

That said, I still think we should wrangle with this one a little more to tidy it up. I am just not sure what that would be yet. 

If you have any more questions about cross wiring, please fire away.

Reply:
> Other than being a little ugly

Yes, it is nasty.

Reply:
> Yes, this would be hard. What CrossWiring is really trying to solve, is making types registered through Windsor available to framework code buried behind IServiceCollection/ServiceProvider logic like the MVC Razor @Inject directive. This is a registration time consideration. I am worried that construction contribution does not consider the registration requirements and anything you can do during resolve to implement this might already be too late.

Not sure I follow why this would be hard, I understand that you are doing late bound components but ComponentModel construction contributors are run directly after construction of the ComponentModel (before `container.Register()` returns), not at resolve time, see [the docs](https://github.com/castleproject/Windsor/blob/master/docs/componentmodel-construction-contributors.md).

The WCF facility uses `ComponentRegistration` to set up services, and after they are registered in the container it'll do a bunch of other stuff. The Startable facility is just the same, you don't have to implement `IStartable` and can use [the `ComponentRegistration` methods](https://github.com/castleproject/Windsor/blob/master/docs/startable-facility.md) and its CMCC adds a start/stop concern if a new component is startable.

Reply:
Ok, let’s try it this way round. What would they contribute to resolves from IServiceCollection/IServiceProvider?

Reply:
I have seen this code and will definitely give it another look. I am just trying to figure how Windsor abstractions will help here. The problem is on the MS side. Please bear with me.